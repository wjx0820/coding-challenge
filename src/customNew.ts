export function customNew<T>(constructor: Function, ...args: any[]): T {
  // 1. 创建一个空对象，继承 constructor 的原型
  const obj = Object.create(constructor.prototype)
  // 2. 将 obj 作为 this ，执行 constructor ，传入参数
  const res = constructor.apply(obj, args)
  // 3. 如果返回的是值类型，那么就返回 obj
  return res instanceof Object ? res : obj
}

// class Foo {
//     // 属性
//     name: string
//     city: string
//     n: number

//     constructor(name: string, n: number) {
//         this.name = name
//         this.city = '北京'
//         this.n = n
//     }

//     getName() {
//         return this.name
//     }
// }

// const f = new Foo('双越', 100)
// // const f = customNew<Foo>(Foo, '双越', 100)
// console.info(f)
// console.info(f.getName())
