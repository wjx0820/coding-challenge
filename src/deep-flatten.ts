export function deepFlatten1(arr: any[]): any[] {
  const res: any[] = []

  arr.forEach((item) => {
    if (Array.isArray(item)) {
      const flatItem = deepFlatten1(item) // 递归
      flatItem.forEach(n => res.push(n))
    }
    else {
      res.push(item)
    }
  })

  return res
}
export function deepFlatten2(arr: any[]): any[] {
  let res: any[] = []

  arr.forEach((item) => {
    res = res.concat(Array.isArray(item) ? deepFlatten2(item) : item)
  })

  return res
}

// const arr = [1, [2, [3, [4, 5]]]]
// console.log(deepFlatten1(arr))
