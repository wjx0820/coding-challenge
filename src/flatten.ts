export function flatten1(arr: any[]): any[] {
  const res: any[] = []
  arr.forEach((item) => {
    if (Array.isArray(item))
      res.push(...item)

    else
      res.push(item)
  })
  return res
}

export function flatten2(arr: any[]): any[] {
  let res: any[] = []

  arr.forEach((item) => {
    res = res.concat(item)
  })

  return res
}

// const arr = [1, [2, [3, [4, 5]]]]
// console.log(deepFlatten(arr))
