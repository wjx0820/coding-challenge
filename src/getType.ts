export function getType(x: any): string {
  const originType = Object.prototype.toString.call(x) // '[object Object]'
  const spaceIndex = originType.indexOf(' ')
  const type = originType.slice(spaceIndex + 1, -1) // 'Object'
  return type.toLowerCase() // 'object'
}
