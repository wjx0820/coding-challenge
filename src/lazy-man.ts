class LazyMan {
  private name: string
  private tasks: Function[] = []
  private next() {
    const task = this.tasks.shift() // 取出当前task的第一个元素
    if (task) task()
  }

  constructor(name: string) {
    this.name = name

    setTimeout(() => {
      this.next()
    })
  }

  eat(food: string) {
    const task = () => {
      console.info(`${this.name}正在吃${food}`)
      this.next()
    }
    this.tasks.push(task)
    return this
  }

  sleep(seconds: number) {
    const task = () => {
      console.info(`${this.name}开始睡觉`)
      setTimeout(() => {
        console.info(`${this.name}睡了${seconds}秒`)
        this.next()
      }, seconds * 1000)
    }
    this.tasks.push(task)
    return this
  }
}

const me = new LazyMan('小明')
me.eat('面包').sleep(2).eat('饭')
