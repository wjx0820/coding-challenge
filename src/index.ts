
// import './dom-traverse'
import './lazy-man'
