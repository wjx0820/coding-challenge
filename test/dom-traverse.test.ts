import type { IWindow } from 'happy-dom'
import { beforeEach, describe, expect, it } from 'vitest'
import { breadthFirstTraverse, depthFirstTraverse1, depthFirstTraverse2 } from '../src/dom-traverse'

declare global {
  interface Window extends IWindow { }
}

describe('深度优先遍历dom', () => {
  beforeEach(async() => {
    document.body.innerHTML = `<div id="box">
      <ul>
        <li>a</li>
        <li>b</li>
      </ul>
    </div>`
    await window.happyDOM.whenAsyncComplete()
    await new Promise(resolve => setTimeout(resolve, 0))
  })
  function getDOM(): HTMLElement | null | undefined {
    return document.getElementById('box')
  }
  it('getDOM', () => {
    expect(getDOM()?.innerHTML).toMatchInlineSnapshot(`
      "
            <ul>
              <li>a</li>
              <li>b</li>
            </ul>
          "
    `)
  })
})
