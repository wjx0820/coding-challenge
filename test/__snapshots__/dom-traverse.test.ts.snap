// Vitest Snapshot v1

exports[`深度优先遍历dom > dfs > a 1`] = `
"
      hello world
      
      
        a
        b
      
    "
`;
